### PR TITLE
#866 adding explanation to DefaultAccessorNamingStrategy and fixing r…

### PR DESCRIPTION
**Pull request by sjaakd**
…esposibillity issue

* added javadoc
* moved the check on the return type of `isPresenceCheckMethod` to `Executables`
* reshuffled the order of methods a little bit (more logical) in `DefaultAccessorNamingStrategy`

### CODE REVIEW

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 1]
```diff
@@ -1974,3 +1974,131 @@ The order in which the selected methods are applied is roughly determined by the
...
+This can be achieved with implementing the SPI `org.mapstruct.ap.spi.AccessorNamingStrategy` as in the following example. Here's an implemented `org.mapstruct.ap.spi.AccessorNamingStrategy`:
+
+.CustomAccessorNamingStrategy
+====
+[source, java, linenums]
+[subs="verbatim,attributes"]
+----
+/**
+ * A custom {@link AccessorNamingStrategy} recognizing getters in the form of {@code property()} and setters in the
+ * form of {@code withProperty(value)}.
```
Comment (agudian) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: :+1:

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 2]
```diff
@@ -1974,3 +1974,131 @@ The order in which the selected methods are applied is roughly determined by the
...
+
+    @Override
+    public String getPropertyName(ExecutableElement getterOrSetterMethod) {
+        String methodName = getterOrSetterMethod.getSimpleName().toString();
+        return Introspector.decapitalize( methodName.startsWith( "with" ) ? methodName.substring(  4 ) : methodName );
+    }
+}
+----
+====
+The `CustomAccessorNamingStrategy` makes use of the `DefaultAccessorNamingStrategy` (also available in mapstruct-processor) and leaves the default behavior to this class. Adding an SPI is requires adding a file named `org.mapstruct.ap.spi.AccessorNamingStrategy` with the content `org.mapstruct.example.CustomAccessorNamingStrategy` (FQN) to the `META-INF/services/` folder. The SPI classes need to be present as class files on the class path when compiling the target project. So, they need to go into a separate library from place were its used (containing the DTO's and the `mapper`).
```
Comment (agudian) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: How about:

The `CustomAccessorNamingStrategy` makes use of the `DefaultAccessorNamingStrategy` (also available in mapstruct-processor) and relies on that class to leave most of the default behaviour unchanged.

To use a custom SPI implementation, it must be located in a seperate `.jar` file together with the file `META-INF/services/org.mapstruct.ap.spi.AccessorNamingStrategy` with the fully qualified name of your custom implementation as content (e.g. `org.mapstruct.example.CustomAccessorNamingStrategy`). That `.jar` file needs to be added to the annotation processor classpath (i.e. add it next to the place where you added the `mapstruct-processor` jar).

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 3]
```diff
@@ -1974,3 +1974,131 @@ The order in which the selected methods are applied is roughly determined by the
...
+
+This as a rudimentary SPI which of course can be improved.
```
Comment (agudian) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: This sentence can safely be removed :). Doesn't add any useful info for the reader and only seperates the very important next sentence from the previous (kinda cryptic) paragraph.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 4]
```diff
@@ -1974,3 +1974,131 @@ The order in which the selected methods are applied is roughly determined by the
...
+
+Fore more details: There's the above example is present in our our examples repository (https://github.com/mapstruct/mapstruct-examples).
```
Comment (agudian) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: "Fore more details: The above example as working code is present in our examples repository (https://github.com/mapstruct/mapstruct-examples)."

I even wonder if we should put that "Hey, check the example repo for the details" stuff into a box to highlight it more..? Or would a user interested in implementing an SPI be disciplined enough to properly read this section to the very end? Hmm... no idea.

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 5]
```diff
@@ -55,7 +55,18 @@ else if ( isPresenceCheckMethod( method ) ) {
...
     }
 
-    private boolean isGetterMethod(ExecutableElement method) {
+    /**
+     * Returns true when the {@link ExecutableElement} is a getter method. A getter is a getter when it starts with
+     * 'get' and the return type is any type other than void OR the getter starts with 'is' and the type returned is
+     * a primitive or the wrapper for boolean {@link Boolean}. NOTE: the latter does strictly not comply to the bean
+     * convention. The remainder of the name is supposed to reflect the property name.
+     *
+     * MapStruct will enforce that there's no argument.
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: Kids, there will be no argument! :neckbeard: 

How about: 
```
    /**
     * Returns @{code true} when the {@link ExecutableElement} is a getter method. A method is a getter when it starts with
     * 'get' and the return type is any type other than {@code void}, OR the getter starts with 'is' and the type returned is
     * a primitive or the wrapper for {@code boolean}. NOTE: the latter does strictly not comply to the bean
     * convention. The remainder of the name is supposed to reflect the property name.
     * <p>
     * The calling code guarantees that the given method has no arguments.
    ...
```

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 6]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
         return isNonBooleanGetterName || ( isBooleanGetterName && returnTypeIsBoolean );
     }
 
+
+    /**
+     * Returns true when the {@link ExecutableElement} is a setter method. A setter starts with 'set'. The remainder of
+     * the name is supposed to reflect the property name.
+     *
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: add `<p>` here as well...

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 7]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
...
+     * MapStruct will ensure that there's only one argument.
+     *
+     * @param method to be analyzed
+     * @return true when the method is a getter.
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: ... a setter.

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 8]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
...
+     */
     public boolean isSetterMethod(ExecutableElement method) {
         String methodName = method.getSimpleName().toString();
 
         return methodName.startsWith( "set" ) && methodName.length() > 3;
     }
 
+    /**
+     * Returns true when the {@link ExecutableElement} is a setter method. An adder method starts with 'add'. The
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: `... is an <em>adder</em> method.`

Do we need to add some documentation about the singularization of adder method names? (property "children", but "addChild")

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 9]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
...
+     * @return true when the method is an adder method.
+     */
     public boolean isAdderMethod(ExecutableElement method) {
         String methodName = method.getSimpleName().toString();
 
         return methodName.startsWith( "add" ) && methodName.length() > 3;
     }
 
+    /**
+     * Returns true when the {@link ExecutableElement} is a setter method. An presence check method starts with
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: `... is a <em>presence check</em> method that checks if the corresponding property is present (e.g. not null, not nil, ..). A presence check method ...`.

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 10]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
...
+     * 'has'. The remainder of the name is supposed to reflect the property name.
+     *
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: `<p>`

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 11]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
...
+     * MapStruct will ensure that there's only one argument and that the return type is boolean or a {@link Boolean}
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: one argument?

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java [hunk 12]
```diff
@@ -68,36 +79,103 @@ private boolean isGetterMethod(ExecutableElement method) {
...
     @Override
     public String getPropertyName(ExecutableElement getterOrSetterMethod) {
         String methodName = getterOrSetterMethod.getSimpleName().toString();
         return Introspector.decapitalize( methodName.substring( methodName.startsWith( "is" ) ? 2 : 3 ) );
     }
 
+    /**
+     * Adder methods are used to add elements to collections on a target bean. A typical use case is JPA. The
+     * convention is that the element name will be equal to the remainder of the add method. Example: 'addElement'
+     * element name will be 'element'.
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: Ah, okay... perhaps add a `{@link ..} to this method above in the isAdderMethod doc.

**Comment by agudian** [comment]
Okay, done with the next round. Sorry that I left off the last time and didn't really help checking the javadocs in the first place.

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java (hunk 5 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: The calling code by itself is also a bit cryptic I guess. Perhaps: MapStruct guarantees that the given method has no arguments.

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java (hunk 8 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: That's a good idea.. I'll add.

File: processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java (hunk 5 above)
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/DefaultAccessorNamingStrategy.java]: Sounds good!

**Comment by sjaakd** [comment]
merged

@agudian  many thanks for the tedious reviewing..

**Pull request closed without merging**